File: internal/agent/fixflow/applyfix.go
package fixflow

import (
	"context"
	"fmt"
	"os"
	"path/filepath"

	"github.com/jkjamies/automation-agent/internal/githubapi"
	"github.com/jkjamies/automation-agent/internal/gitrepo"
)

// GitHub is the slice of githubapi the apply step needs (consumer-defined, fakeable).
type GitHub interface {
	FindAgentPRs(ctx context.Context, owner, repo, label string) ([]githubapi.PR, error)
	CreatePR(ctx context.Context, owner, repo string, in githubapi.PRInput) (githubapi.PR, error)
	AddLabels(ctx context.Context, owner, repo string, number int, labels ...string) error
}

// FileEdit is a whole-file write an analyze step produces (a rewritten source file,
// a generated test file, …).
type FileEdit struct {
	Path    string // repo-relative path
	Content string
}

// ApplyConfig parameterizes one apply.
type ApplyConfig struct {
	Owner, Repo   string
	CloneURL      string
	Token         string
	Base          string // base branch the PR targets
	Branch        string // agent working branch
	NewBranch     bool   // true on kickoff (create from base); false on retry (reuse remote branch)
	Label         string
	CommitMessage string
	PRTitle       string
	PRBody        string
	Author        gitrepo.Author
}

// ApplyResult is the outcome of one apply.
type ApplyResult struct {
	PR      githubapi.PR
	HeadSHA string
}

// Open clones the repo into a fresh temp dir and checks out the agent branch — the
// single checkout the explorer reads, the executor writes into, and the commit step
// pushes. NewBranch=true creates the branch from HEAD (kickoff); false checks out the
// existing remote branch (retry). The caller must os.RemoveAll(repo.Dir()) when done.
func Open(ctx context.Context, cfg ApplyConfig) (*gitrepo.Repo, error) {
	dir, err := os.MkdirTemp("", "agentfix-*")
	if err != nil {
		return nil, fmt.Errorf("tempdir: %w", err)
	}
	repo, err := gitrepo.Clone(ctx, cfg.CloneURL, dir, cfg.Token)
	if err != nil {
		os.RemoveAll(dir)
		return nil, err
	}
	if cfg.NewBranch {
		err = repo.Checkout(cfg.Branch, true)
	} else {
		err = repo.CheckoutRemote(cfg.Branch)
	}
	if err != nil {
		os.RemoveAll(dir)
		return nil, err
	}
	return repo, nil
}

// Commit writes edits into the working tree, commits, pushes, and ensures a labeled
// PR exists.
func Commit(ctx context.Context, gh GitHub, repo *gitrepo.Repo, cfg ApplyConfig, edits []FileEdit) (ApplyResult, error) {
	if len(edits) == 0 {
		return ApplyResult{}, fmt.Errorf("apply: no edits to apply")
	}
	if err := writeEdits(repo, edits); err != nil {
		return ApplyResult{}, err
	}
	sha, err := repo.CommitAll(cfg.CommitMessage, cfg.Author)
	if err != nil {
		return ApplyResult{}, err
	}
	if err := repo.Push(ctx); err != nil {
		return ApplyResult{}, err
	}
	pr, err := ensurePR(ctx, gh, cfg)
	if err != nil {
		return ApplyResult{}, err
	}
	return ApplyResult{PR: pr, HeadSHA: sha}, nil
}

// ApplyFix opens a checkout and commits edits in one step (no analysis in between) —
// a convenience used in tests; the engine interleaves analysis between Open and Commit.
func ApplyFix(ctx context.Context, gh GitHub, cfg ApplyConfig, edits []FileEdit) (ApplyResult, error) {
	repo, err := Open(ctx, cfg)
	if err != nil {
		return ApplyResult{}, err
	}
	defer os.RemoveAll(repo.Dir())
	return Commit(ctx, gh, repo, cfg, edits)
}

func writeEdits(repo *gitrepo.Repo, edits []FileEdit) error {
	for _, e := range edits {
		full := repo.Path(e.Path)
		if err := os.MkdirAll(filepath.Dir(full), 0o755); err != nil {
			return fmt.Errorf("mkdir for %s: %w", e.Path, err)
		}
		if err := os.WriteFile(full, []byte(e.Content), 0o644); err != nil {
			return fmt.Errorf("write %s: %w", e.Path, err)
		}
	}
	return nil
}

// ensurePR returns the existing agent PR for the branch, or creates and labels one.
func ensurePR(ctx context.Context, gh GitHub, cfg ApplyConfig) (githubapi.PR, error) {
	existing, err := gh.FindAgentPRs(ctx, cfg.Owner, cfg.Repo, cfg.Label)
	if err != nil {
		return githubapi.PR{}, err
	}
	for _, pr := range existing {
		if pr.Branch == cfg.Branch {
			return pr, nil
		}
	}
	pr, err := gh.CreatePR(ctx, cfg.Owner, cfg.Repo, githubapi.PRInput{
		Title: cfg.PRTitle, Head: cfg.Branch, Base: cfg.Base, Body: cfg.PRBody,
	})
	if err != nil {
		return githubapi.PR{}, err
	}
	if err := gh.AddLabels(ctx, cfg.Owner, cfg.Repo, pr.Number, cfg.Label); err != nil {
		return githubapi.PR{}, err
	}
	pr.Labels = append(pr.Labels, cfg.Label)
	return pr, nil
}

File: kotlin/src/main/kotlin/com/automation/agent/agent/fixflow/ApplyFix.kt
package com.automation.agent.agent.fixflow

import com.automation.agent.githubapi.Comparison
import com.automation.agent.githubapi.Pr
import com.automation.agent.githubapi.PrInput
import com.automation.agent.gitrepo.Author
import com.automation.agent.gitrepo.Repo
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.nio.file.Files

/** The slice of `githubapi` the apply step needs (consumer-defined, fakeable). */
interface GitHub {
    suspend fun findOpenPrByBranch(owner: String, repo: String, branch: String): Pr?
    suspend fun createPr(owner: String, repo: String, input: PrInput): Pr
    suspend fun addLabels(owner: String, repo: String, number: Int, labels: List<String>)

    /** The base...head comparison for a terminal summary (what the agent changed on the PR). */
    suspend fun compare(owner: String, repo: String, base: String, head: String): Comparison
}

/** A whole-file write an analyze step produces (a rewritten source file, a generated test, …). */
data class FileEdit(val path: String, val content: String)

/** Parameterizes one apply. */
data class ApplyConfig(
    val owner: String,
    val repo: String,
    val cloneUrl: String,
    val token: String = "",
    /**
     * Explicit private-key path for an ssh [cloneUrl] (GIT_SSH_KEY); empty falls back to ssh-agent
     * then the default identity files. Ignored for an https [cloneUrl].
     */
    val sshKey: String = "",
    val base: String, // base branch the PR targets
    val branch: String, // agent working branch
    val newBranch: Boolean, // true on kickoff (create from base); false on retry (reuse remote branch)
    val label: String,
    val commitMessage: String,
    val prTitle: String,
    val prBody: String,
    val author: Author,
)

/** The outcome of one apply. */
data class ApplyResult(val pr: Pr, val headSha: String)

/**
 * Clones the repo into a fresh temp dir and checks out the agent branch — the single checkout the
 * explorer reads, the executor writes into, and the commit step pushes. [ApplyConfig.newBranch]
 * creates the branch from HEAD (kickoff); false checks out the existing remote branch (retry). The
 * caller must delete `repo.dir()` when done.
 */
suspend fun open(cfg: ApplyConfig): Repo {
    val dir = withContext(Dispatchers.IO) { Files.createTempDirectory("agentfix-").toFile() }
    val repo =
        try {
            Repo.clone(cfg.cloneUrl, dir.path, cfg.token, cfg.sshKey)
        } catch (e: Throwable) {
            withContext(Dispatchers.IO) { dir.deleteRecursively() }
            throw e
        }
    try {
        if (cfg.newBranch) repo.checkout(cfg.branch, create = true) else repo.checkoutRemote(cfg.branch)
    } catch (e: Throwable) {
        withContext(Dispatchers.IO) { dir.deleteRecursively() }
        throw e
    }
    return repo
}

/** Writes edits into the working tree, commits, pushes, and ensures a labeled PR exists. */
suspend fun commit(gh: GitHub, repo: Repo, cfg: ApplyConfig, edits: List<FileEdit>): ApplyResult {
    require(edits.isNotEmpty()) { "apply: no edits to apply" }
    writeEdits(repo, edits)
    val sha = repo.commitAll(cfg.commitMessage, cfg.author)
    repo.push()
    val pr = ensurePR(gh, cfg)
    return ApplyResult(pr = pr, headSha = sha)
}

/**
 * Opens a checkout and commits edits in one step (no analysis in between) — a convenience used in
 * tests; the engine interleaves analysis between [open] and [commit].
 */
suspend fun applyFix(gh: GitHub, cfg: ApplyConfig, edits: List<FileEdit>): ApplyResult {
    val repo = open(cfg)
    return try {
        commit(gh, repo, cfg, edits)
    } finally {
        // Release the JGit handles before deleting the checkout (see Engine.attemptOnce).
        repo.close()
        withContext(Dispatchers.IO) { File(repo.dir()).deleteRecursively() }
    }
}

private suspend fun writeEdits(repo: Repo, edits: List<FileEdit>) = withContext(Dispatchers.IO) {
    for (edit in edits) {
        // Reject LLM-controlled paths that escape the checkout (path traversal).
        val full =
            try {
                safeJoin(repo.dir(), edit.path)
            } catch (e: IllegalArgumentException) {
                throw IllegalArgumentException("reject edit \"${edit.path}\": ${e.message}")
            }
        File(full).parentFile?.mkdirs()
        File(full).writeText(edit.content)
    }
}

/**
 * Returns the existing open PR for the branch, or creates and labels one. Lookup is by head branch
 * (not the agent label, which is write-only and never read back).
 */
private suspend fun ensurePR(gh: GitHub, cfg: ApplyConfig): Pr {
    gh.findOpenPrByBranch(cfg.owner, cfg.repo, cfg.branch)?.let { return it }
    val pr = gh.createPr(cfg.owner, cfg.repo, PrInput(title = cfg.prTitle, head = cfg.branch, base = cfg.base, body = cfg.prBody))
    gh.addLabels(cfg.owner, cfg.repo, pr.number, listOf(cfg.label))
    return pr.copy(labels = pr.labels + cfg.label)
}
